File: android/src/main/java/com/eugeniuszx/reactnativemediapicker/ImageProcessor.kt
package com.eugeniuszx.reactnativemediapicker

import android.content.ContentResolver
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import android.os.Build
import android.util.Base64
import androidx.exifinterface.media.ExifInterface
import java.io.File
import java.io.FileOutputStream

internal class ImageProcessor(
  private val resolver: ContentResolver,
  private val tempFiles: TempFileStore,
) {
  fun process(
    uri: Uri,
    format: RequestedFormat,
    maxWidth: Int,
    maxHeight: Int,
    quality: Int,
    includeBase64: Boolean,
    stripMetadata: Boolean = false,
    includeExif: Boolean = false,
    suggestedName: String? = null,
  ): AssetPayload {
    val srcMime = MediaFormat.normalizeMime(resolver.getType(uri))
    val bounds = readBounds(uri)
    val orientation = readOrientation(uri)
    val exif = if (includeExif) ExifReader.read(resolver, uri) else null
    val isAnimated = srcMime == "image/gif" ||
      (srcMime == "image/webp" && isAnimatedWebp(uri))

    val output = OutputPlan.resolve(srcMime, format, isAnimated)
    val plan = DecodePlan.compute(
      pixelWidth = bounds.first,
      pixelHeight = bounds.second,
      orientation = orientation,
      maxWidth = maxWidth,
      maxHeight = maxHeight,
      isAnimated = output.preserveAnimation,
    )

    val willTransform = plan.needsTransform || output.forceReencode
    val action = MetadataPlan.resolve(
      stripMetadata = stripMetadata,
      willTransform = willTransform,
      preserveSource = MetadataPlan.preservesSource(srcMime, output.preserveAnimation),
      canScrub = MetadataPlan.canScrub(srcMime),
    )

    if (willTransform || action == MetadataAction.FORCE_REENCODE) {
      return transform(
        uri, output.target, plan, orientation, quality, includeBase64, suggestedName, exif,
      )
    }

    // A failed scrub falls through to a re-encode, so a strip request is never silently ignored.
    return passthrough(
      uri, srcMime, plan, includeBase64, suggestedName,
      scrub = action == MetadataAction.SCRUB, exif = exif,
    ) ?: transform(
      uri, output.target, plan, orientation, quality, includeBase64, suggestedName, exif,
    )
  }

  private fun readBounds(uri: Uri): Pair<Int, Int> {
    val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    resolver.openInputStream(uri).use { BitmapFactory.decodeStream(it, null, options) }
    return options.outWidth to options.outHeight
  }

  private fun readOrientation(uri: Uri): ExifOrientation {
    val value = try {
      resolver.openInputStream(uri).use { stream ->
        stream ?: return ExifOrientation.UPRIGHT
        ExifInterface(stream).getAttributeInt(
          ExifInterface.TAG_ORIENTATION,
          ExifInterface.ORIENTATION_NORMAL,
        )
      }
    } catch (e: Exception) {
      ExifInterface.ORIENTATION_NORMAL
    }
    return ExifOrientation.fromExifValue(value)
  }

  private fun isAnimatedWebp(uri: Uri): Boolean = try {
    resolver.openInputStream(uri).use { stream ->
      stream ?: return false
      val header = ByteArray(21)
      var total = 0
      while (total < header.size) {
        val read = stream.read(header, total, header.size - total)
        if (read == -1) break
        total += read
      }
      total >= header.size && MediaFormat.isAnimatedWebp(header)
    }
  } catch (e: Exception) {
    false
  }

  private fun passthrough(
    uri: Uri,
    mime: String,
    plan: DecodePlan,
    includeBase64: Boolean,
    suggestedName: String?,
    scrub: Boolean,
    exif: ExifPayload?,
  ): AssetPayload? {
    val outFile = tempFiles.createFile(MediaFormat.extensionForMime(mime))
    resolver.openInputStream(uri).use { input ->
      input ?: throw IllegalStateException("Failed to open image stream")
      FileOutputStream(outFile).use { output -> input.copyTo(output) }
    }

    // A source carrying an XMP packet is declined before the scrub is even attempted:
    // ExifInterface cannot remove XMP, and a packet can carry `exif:GPSLatitude`, `tiff:Make` and
    // `tiff:Model` of its own — so a scrubbed file would still leak exactly what the caller asked
    // to remove. Returning null sends it down the re-encode path instead, which produces a
    // metadata-free file by construction. The re-encode is the deliberate price of a guarantee
    // that holds for every input rather than one that quietly leaks for files from an
    // XMP-writing pipeline (Lightroom, Google Photos).
    if (scrub && (XMPPacket.isPresent(outFile) || !MetadataScrubber.scrub(outFile))) {
      outFile.delete()
      return null
    }

    val base64 = if (includeBase64) {
      Base64.encodeToString(outFile.readBytes(), Base64.NO_WRAP)
    } else {
      null
    }
    return payload(
      outFile, mime, plan.displayWidth, plan.displayHeight, base64, suggestedName, exif,
    )
  }

  private fun transform(
    uri: Uri,
    target: MediaFormat.OutputFormat,
    plan: DecodePlan,
    orientation: ExifOrientation,
    quality: Int,
    includeBase64: Boolean,
    suggestedName: String?,
    exif: ExifPayload?,
  ): AssetPayload {
    val decodeOptions = BitmapFactory.Options().apply { inSampleSize = plan.sampleSize }
    var bitmap = resolver.openInputStream(uri).use {
      BitmapFactory.decodeStream(it, null, decodeOptions)
    } ?: throw IllegalStateException("Failed to decode image")

    bitmap = applyOrientation(bitmap, orientation)
    bitmap = scaleTo(bitmap, plan.targetWidth, plan.targetHeight)

    val outMime = MediaFormat.reencodeMime(target)
    val outFile = tempFiles.createFile(MediaFormat.extensionForMime(outMime))
    FileOutputStream(outFile).use { output ->
      if (!bitmap.compress(compressFormat(target), quality, output)) {
        bitmap.recycle()
        outFile.delete()
        throw IllegalStateException("Failed to encode image as $outMime")
      }
    }

    val width = bitmap.width
    val height = bitmap.height
    bitmap.recycle()

    val base64 = if (includeBase64) {
      Base64.encodeToString(outFile.readBytes(), Base64.NO_WRAP)
    } else {
      null
    }
    return payload(outFile, outMime, width, height, base64, suggestedName, exif)
  }

  private fun applyOrientation(bitmap: Bitmap, orientation: ExifOrientation): Bitmap {
    if (orientation == ExifOrientation.UPRIGHT) return bitmap
    val matrix = Matrix().apply {
      if (orientation.rotationDegrees != 0) postRotate(orientation.rotationDegrees.toFloat())
      if (orientation.isMirrored) postScale(-1f, 1f)
    }
    val oriented =
      Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    if (oriented != bitmap) bitmap.recycle()
    return oriented
  }

  private fun scaleTo(bitmap: Bitmap, width: Int, height: Int): Bitmap {
    if (bitmap.width == width && bitmap.height == height) return bitmap
    val scaled = Bitmap.createScaledBitmap(bitmap, width, height, true)
    if (scaled != bitmap) bitmap.recycle()
    return scaled
  }

  @Suppress("DEPRECATION")
  private fun compressFormat(format: MediaFormat.OutputFormat): Bitmap.CompressFormat =
    when (format) {
      MediaFormat.OutputFormat.PNG -> Bitmap.CompressFormat.PNG
      MediaFormat.OutputFormat.JPEG -> Bitmap.CompressFormat.JPEG
      MediaFormat.OutputFormat.WEBP ->
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
          Bitmap.CompressFormat.WEBP_LOSSY
        } else {
          Bitmap.CompressFormat.WEBP
        }
    }

  private fun payload(
    file: File,
    mime: String,
    width: Int,
    height: Int,
    base64: String?,
    suggestedName: String?,
    exif: ExifPayload?,
  ) = AssetPayload(
    uri = Uri.fromFile(file).toString(),
    mime = mime,
    fileName = AssetFileName.resolve(suggestedName, file.name, file.extension),
    fileSize = file.length(),
    width = width,
    height = height,
    base64 = base64,
    exif = exif,
  )
}

File: ios/ImageProcessor.swift
import CoreGraphics
import ImageIO
import UIKit
import UniformTypeIdentifiers

struct ImageProcessor {
  private let tempFiles: TempFileStore

  init(tempFiles: TempFileStore) {
    self.tempFiles = tempFiles
  }

  func process(
    data: Data,
    format: ImageFormat,
    requested: RequestedFormat,
    maxWidth: Int,
    maxHeight: Int,
    quality: Double,
    includeBase64: Bool,
    stripMetadata: Bool = false,
    includeExif: Bool = false,
    suggestedName: String? = nil
  ) -> AssetPayload? {
    let metadata = readMetadata(from: data)
    let exif = includeExif ? ExifReader.read(from: data) : nil
    let isAnimated =
      format == .gif
      || (format == .webp && ImageFormat.isAnimatedWebP(header: data))
    let output = OutputPlan.resolve(
      source: format,
      requested: requested,
      isAnimatedSource: isAnimated
    )
    let plan = DecodePlan.compute(
      pixelWidth: metadata.pixelWidth,
      pixelHeight: metadata.pixelHeight,
      orientation: metadata.orientation,
      maxWidth: maxWidth,
      maxHeight: maxHeight,
      isAnimated: output.preserveAnimation
    )

    let willTransform = plan.needsTransform || output.forceReencode
    let action = MetadataPlan.resolve(
      stripMetadata: stripMetadata,
      willTransform: willTransform,
      preserveSource: output.preserveAnimation || format == .gif,
      canScrub: MetadataPlan.canScrub(format)
    )

    if !willTransform, action != .forceReencode {
      // `.scrub` falls through to the transform path when ImageIO refuses the rewrite, so a
      // strip request is never silently ignored.
      let payloadData = action == .scrub ? MetadataScrubber.scrubbed(data: data) : data
      if let payloadData {
        return write(
          data: payloadData,
          format: format,
          width: plan.displayWidth,
          height: plan.displayHeight,
          includeBase64: includeBase64,
          suggestedName: suggestedName,
          exif: exif
        )
      }
    }

    guard let image = UIImage(data: data) else { return nil }
    let targetWidth = plan.targetWidth > 0 ? plan.targetWidth : Int(image.size.width * image.scale)
    let targetHeight =
      plan.targetHeight > 0 ? plan.targetHeight : Int(image.size.height * image.scale)
    let resized = redraw(image, width: targetWidth, height: targetHeight)
    return encodeAndWrite(
      resized,
      format: output.target,
      quality: quality,
      includeBase64: includeBase64,
      suggestedName: suggestedName,
      exif: exif
    )
  }

  func process(
    capturedImage image: UIImage,
    requested: RequestedFormat,
    maxWidth: Int,
    maxHeight: Int,
    quality: Double,
    includeBase64: Bool,
    exif: ExifPayload? = nil
  ) -> AssetPayload? {
    let pixelWidth = image.cgImage?.width ?? Int(image.size.width * image.scale)
    let pixelHeight = image.cgImage?.height ?? Int(image.size.height * image.scale)
    let output = OutputPlan.resolve(
      source: .jpeg,
      requested: requested,
      isAnimatedSource: false
    )
    let plan = DecodePlan.compute(
      pixelWidth: pixelWidth,
      pixelHeight: pixelHeight,
      orientation: orientation(of: image),
      maxWidth: maxWidth,
      maxHeight: maxHeight,
      isAnimated: false
    )
    let resized =
      plan.needsTransform
      ? redraw(image, width: plan.targetWidth, height: plan.targetHeight)
      : redraw(image, width: plan.displayWidth, height: plan.displayHeight)
    return encodeAndWrite(
      resized,
      format: output.target,
      quality: quality,
      includeBase64: includeBase64,
      exif: exif
    )
  }

  private struct Metadata {
    let pixelWidth: Int
    let pixelHeight: Int
    let orientation: ImageOrientation
  }

  private func readMetadata(from data: Data) -> Metadata {
    guard let source = CGImageSourceCreateWithData(data as CFData, nil),
      let props = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any]
    else {
      NSLog("[ReactNativeMediaPicker] warning: could not read image metadata")
      return Metadata(pixelWidth: 0, pixelHeight: 0, orientation: .upright)
    }
    let width = (props[kCGImagePropertyPixelWidth] as? Int) ?? 0
    let height = (props[kCGImagePropertyPixelHeight] as? Int) ?? 0
    let exif = (props[kCGImagePropertyOrientation] as? NSNumber)?.intValue ?? 1
    return Metadata(
      pixelWidth: width,
      pixelHeight: height,
      orientation: ImageOrientation.from(exifValue: exif)
    )
  }

  private func orientation(of image: UIImage) -> ImageOrientation {
    switch image.imageOrientation {
    case .up: return .upright
    case .upMirrored: return ImageOrientation.from(exifValue: 2)
    case .down: return ImageOrientation.from(exifValue: 3)
    case .downMirrored: return ImageOrientation.from(exifValue: 4)
    case .leftMirrored: return ImageOrientation.from(exifValue: 5)
    case .right: return ImageOrientation.from(exifValue: 6)
    case .rightMirrored: return ImageOrientation.from(exifValue: 7)
    case .left: return ImageOrientation.from(exifValue: 8)
    @unknown default: return .upright
    }
  }

  private func redraw(_ image: UIImage, width: Int, height: Int) -> UIImage {
    let size = CGSize(width: max(1, width), height: max(1, height))
    let format = UIGraphicsImageRendererFormat.default()
    format.scale = 1
    format.opaque = false
    return UIGraphicsImageRenderer(size: size, format: format).image { _ in
      image.draw(in: CGRect(origin: .zero, size: size))
    }
  }

  private func encode(
    _ image: UIImage,
    format: ImageFormat,
    quality: Double
  ) -> (data: Data, format: ImageFormat)? {
    switch format {
    case .png:
      guard let data = image.pngData() else { return nil }
      return (data, .png)
    case .heic:
      if let data = heicData(from: image, quality: quality) {
        return (data, .heic)
      }
      guard let data = image.jpegData(compressionQuality: CGFloat(quality)) else { return nil }
      return (data, .jpeg)
    case .jpeg, .gif, .webp:
      guard let data = image.jpegData(compressionQuality: CGFloat(quality)) else { return nil }
      return (data, .jpeg)
    }
  }

  private func heicData(from image: UIImage, quality: Double) -> Data? {
    guard let cgImage = image.cgImage else { return nil }
    let output = NSMutableData()
    guard
      let destination = CGImageDestinationCreateWithData(
        output, UTType.heic.identifier as CFString, 1, nil
      )
    else { return nil }
    CGImageDestinationAddImage(
      destination,
      cgImage,
      [kCGImageDestinationLossyCompressionQuality: quality] as CFDictionary
    )
    guard CGImageDestinationFinalize(destination) else { return nil }
    return output as Data
  }

  private func encodeAndWrite(
    _ image: UIImage,
    format: ImageFormat,
    quality: Double,
    includeBase64: Bool,
    suggestedName: String? = nil,
    exif: ExifPayload?
  ) -> AssetPayload? {
    guard let encoded = encode(image, format: format, quality: quality) else { return nil }
    let cgImage = image.cgImage
    let pixelWidth = cgImage?.width ?? Int(image.size.width)
    let pixelHeight = cgImage?.height ?? Int(image.size.height)
    return write(
      data: encoded.data,
      format: encoded.format,
      width: pixelWidth,
      height: pixelHeight,
      includeBase64: includeBase64,
      suggestedName: suggestedName,
      exif: exif
    )
  }

  private func write(
    data: Data,
    format: ImageFormat,
    width: Int,
    height: Int,
    includeBase64: Bool,
    suggestedName: String?,
    exif: ExifPayload?
  ) -> AssetPayload? {
    do {
      let url = try tempFiles.makeFileURL(fileExtension: format.fileExtension)
      try data.write(to: url)
      return AssetPayload(
        uri: url.absoluteString,
        mime: format.mime,
        fileName: AssetFileName.resolve(
          suggested: suggestedName,
          fallback: url.lastPathComponent,
          fileExtension: url.pathExtension
        ),
        fileSize: data.count,
        width: width,
        height: height,
        base64: includeBase64 ? data.base64EncodedString() : nil,
        exif: exif
      )
    } catch {
      NSLog("[ReactNativeMediaPicker] failed to write asset: %@", error.localizedDescription)
      return nil
    }
  }
}
